File: android/app/src/main/java/my/ssdid/wallet/feature/history/TxHistoryScreen.kt
package my.ssdid.wallet.feature.history

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.hilt.navigation.compose.hiltViewModel
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import my.ssdid.wallet.domain.history.ActivityRepository
import my.ssdid.wallet.domain.model.ActivityRecord
import my.ssdid.wallet.domain.model.ActivityType
import my.ssdid.wallet.ui.theme.*
import javax.inject.Inject

private data class ActivityStyle(val icon: String, val color: Color, val dimColor: Color)

private fun ActivityType.style(): ActivityStyle = when (this) {
    ActivityType.IDENTITY_CREATED -> ActivityStyle("\u2B21", Accent, AccentDim)
    ActivityType.SERVICE_REGISTERED -> ActivityStyle("\uD83D\uDD17", Success, SuccessDim)
    ActivityType.AUTHENTICATED -> ActivityStyle("\u2713", Classical, ClassicalDim)
    ActivityType.TX_SIGNED -> ActivityStyle("\u270E", Warning, WarningDim)
    ActivityType.KEY_ROTATED -> ActivityStyle("\uD83D\uDD04", Pqc, PqcDim)
    ActivityType.CREDENTIAL_RECEIVED -> ActivityStyle("\uD83D\uDCC4", Success, SuccessDim)
    ActivityType.BACKUP_CREATED -> ActivityStyle("\uD83D\uDCBE", Accent, AccentDim)
    else -> ActivityStyle("\u2022", TextSecondary, BgCard)
}

@HiltViewModel
class TxHistoryViewModel @Inject constructor(
    private val activityRepository: ActivityRepository
) : ViewModel() {
    private val _activities = MutableStateFlow<List<ActivityRecord>>(emptyList())
    val activities = _activities.asStateFlow()

    init {
        viewModelScope.launch {
            _activities.value = activityRepository.listActivities()
        }
    }
}

@Composable
fun TxHistoryScreen(
    onBack: () -> Unit,
    viewModel: TxHistoryViewModel = hiltViewModel()
) {
    val activities by viewModel.activities.collectAsState()
    val groupedItems = activities.groupBy { it.timestamp.take(10) }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(BgPrimary)
            .statusBarsPadding()
    ) {
        // Header
        Row(Modifier.padding(20.dp), verticalAlignment = Alignment.CenterVertically) {
            TextButton(onClick = onBack) { Text("\u2190", color = TextPrimary, fontSize = 20.sp) }
            Spacer(Modifier.width(12.dp))
            Text("Activity History", style = MaterialTheme.typography.titleLarge)
        }

        if (activities.isEmpty()) {
            Box(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(32.dp),
                contentAlignment = Alignment.Center
            ) {
                Column(horizontalAlignment = Alignment.CenterHorizontally) {
                    Box(
                        modifier = Modifier
                            .size(72.dp)
                            .clip(RoundedCornerShape(20.dp))
                            .background(BgCard),
                        contentAlignment = Alignment.Center
                    ) {
                        Text("\uD83D\uDCCB", fontSize = 32.sp)
                    }
                    Spacer(Modifier.height(16.dp))
                    Text("No activity yet", style = MaterialTheme.typography.headlineSmall, color = TextPrimary)
                    Spacer(Modifier.height(4.dp))
                    Text("Your identity and transaction history will appear here.", color = TextSecondary, fontSize = 14.sp)
                }
            }
        } else {
            LazyColumn(
                Modifier.padding(horizontal = 20.dp),
                verticalArrangement = Arrangement.spacedBy(6.dp)
            ) {
                groupedItems.forEach { (date, items) ->
                    item {
                        Text(
                            date.uppercase(),
                            style = MaterialTheme.typography.labelMedium,
                            modifier = Modifier.padding(top = 8.dp, bottom = 6.dp)
                        )
                    }
                    items(items) { record ->
                        val style = record.type.style()
                        val title = record.type.name
                            .replace("_", " ")
                            .lowercase()
                            .replaceFirstChar { it.uppercase() }
                        val subtitle = record.serviceUrl ?: record.did
                        val time = if (record.timestamp.length > 11) {
                            record.timestamp.substring(11).take(5)
                        } else {
                            ""
                        }

                        Card(
                            Modifier.fillMaxWidth(),
                            shape = RoundedCornerShape(12.dp),
                            colors = CardDefaults.cardColors(containerColor = BgCard)
                        ) {
                            Row(
                                Modifier.padding(14.dp),
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                // Type icon
                                Box(
                                    modifier = Modifier
                                        .size(40.dp)
                                        .clip(RoundedCornerShape(12.dp))
                                        .background(style.dimColor),
                                    contentAlignment = Alignment.Center
                                ) {
                                    Text(style.icon, fontSize = 16.sp, color = style.color)
                                }
                                Spacer(Modifier.width(12.dp))
                                Column(Modifier.weight(1f)) {
                                    Text(
                                        title,
                                        fontSize = 14.sp,
                                        fontWeight = FontWeight.Medium,
                                        color = TextPrimary
                                    )
                                    Text(
                                        subtitle,
                                        fontSize = 12.sp,
                                        color = TextTertiary
                                    )
                                }
                                Text(
                                    time,
                                    fontSize = 12.sp,
                                    color = TextTertiary
                                )
                            }
                        }
                    }
                }

                // Bottom spacer
                item { Spacer(Modifier.height(20.dp)) }
            }
        }
    }
}

File: ios/SsdidWallet/Feature/History/TxHistoryScreen.swift
import SwiftUI
import SsdidCore

struct TxHistoryScreen: View {
    @Environment(AppRouter.self) private var router

    @State private var activities: [ActivityRecord] = []

    private var groupedActivities: [(date: String, records: [ActivityRecord])] {
        let grouped = Dictionary(grouping: activities) { String($0.timestamp.prefix(10)) }
        return grouped.sorted { $0.key > $1.key }.map { (date: $0.key, records: $0.value) }
    }

    var body: some View {
        VStack(spacing: 0) {
            // Header
            HStack(spacing: 4) {
                Button { router.pop() } label: {
                    Image(systemName: "chevron.left")
                        .foregroundStyle(Color.textPrimary)
                        .font(.system(size: 20))
                }
                .padding(.leading, 8)

                Text("Activity History")
                    .font(.ssdidHeadline)
                    .foregroundStyle(Color.textPrimary)

                Spacer()
            }
            .padding(.vertical, 12)
            .padding(.trailing, 20)

            if activities.isEmpty {
                Spacer()
                VStack(spacing: 0) {
                    ZStack {
                        RoundedRectangle(cornerRadius: 20)
                            .fill(Color.bgCard)
                            .frame(width: 72, height: 72)
                        Image(systemName: "list.clipboard")
                            .font(.system(size: 32))
                            .foregroundStyle(Color.textTertiary)
                    }
                    Spacer().frame(height: 16)
                    Text("No activity yet")
                        .font(.ssdidHeadline)
                        .foregroundStyle(Color.textPrimary)
                    Spacer().frame(height: 4)
                    Text("Your identity and transaction history will appear here.")
                        .font(.system(size: 14))
                        .foregroundStyle(Color.textSecondary)
                        .multilineTextAlignment(.center)
                }
                .padding(.horizontal, 32)
                Spacer()
            } else {
                ScrollView {
                    LazyVStack(spacing: 6) {
                        ForEach(groupedActivities, id: \.date) { group in
                            Text(group.date.uppercased())
                                .font(.ssdidCaption)
                                .foregroundStyle(Color.textTertiary)
                                .frame(maxWidth: .infinity, alignment: .leading)
                                .padding(.top, 8)
                                .padding(.bottom, 6)

                            ForEach(group.records) { record in
                                activityRow(record)
                            }
                        }

                        Spacer().frame(height: 20)
                    }
                    .padding(.horizontal, 20)
                }
            }
        }
        .background(Color.bgPrimary)
    }

    @ViewBuilder
    private func activityRow(_ record: ActivityRecord) -> some View {
        let style = activityStyle(record.type)
        let title = record.type.rawValue
            .replacingOccurrences(of: "_", with: " ")
            .lowercased()
            .prefix(1).uppercased() + record.type.rawValue
            .replacingOccurrences(of: "_", with: " ")
            .lowercased()
            .dropFirst()
        let subtitle = record.serviceUrl ?? record.did
        let time: String = {
            if record.timestamp.count > 11 {
                return String(record.timestamp.dropFirst(11).prefix(5))
            }
            return ""
        }()

        HStack(spacing: 12) {
            ZStack {
                RoundedRectangle(cornerRadius: 12)
                    .fill(style.dimColor)
                    .frame(width: 40, height: 40)
                Image(systemName: style.icon)
                    .font(.system(size: 16))
                    .foregroundStyle(style.color)
            }
            VStack(alignment: .leading, spacing: 2) {
                Text(title)
                    .font(.system(size: 14, weight: .medium))
                    .foregroundStyle(Color.textPrimary)
                Text(subtitle)
                    .font(.system(size: 12))
                    .foregroundStyle(Color.textTertiary)
                    .lineLimit(1)
            }
            Spacer()
            Text(time)
                .font(.system(size: 12))
                .foregroundStyle(Color.textTertiary)
        }
        .padding(14)
        .background(Color.bgCard)
        .cornerRadius(12)
    }

    private struct ActivityStyle {
        let icon: String
        let color: Color
        let dimColor: Color
    }

    private func activityStyle(_ type: ActivityType) -> ActivityStyle {
        switch type {
        case .IDENTITY_CREATED:
            return ActivityStyle(icon: "hexagon", color: .ssdidAccent, dimColor: .accentDim)
        case .IDENTITY_DEACTIVATED:
            return ActivityStyle(icon: "xmark.hexagon", color: .danger, dimColor: .dangerDim)
        case .SERVICE_REGISTERED:
            return ActivityStyle(icon: "link", color: .success, dimColor: .successDim)
        case .AUTHENTICATED:
            return ActivityStyle(icon: "checkmark.shield", color: .classical, dimColor: .classicalDim)
        case .TX_SIGNED:
            return ActivityStyle(icon: "pencil", color: .warning, dimColor: .warningDim)
        case .KEY_ROTATED:
            return ActivityStyle(icon: "arrow.triangle.2.circlepath", color: .pqc, dimColor: .pqcDim)
        case .CREDENTIAL_RECEIVED:
            return ActivityStyle(icon: "doc.text", color: .success, dimColor: .successDim)
        case .BACKUP_CREATED:
            return ActivityStyle(icon: "externaldrive", color: .ssdidAccent, dimColor: .accentDim)
        case .DEVICE_ENROLLED:
            return ActivityStyle(icon: "iphone", color: .ssdidAccent, dimColor: .accentDim)
        case .DEVICE_REMOVED:
            return ActivityStyle(icon: "iphone.slash", color: .danger, dimColor: .dangerDim)
        case .CREDENTIAL_PRESENTED:
            return ActivityStyle(icon: "doc.badge.arrow.up", color: .classical, dimColor: .classicalDim)
        }
    }
}
